### ISSUE TITLE
DistanceDependentProbabilityConnector cannot be instantiated 

### ISSUE THREAD

**Issue by uahic** [SpiNNakerManchester/sPyNNaker#402]
While instantiation of the DistanceDependentProbabilityConnector the following error occurs

```
/home/schulze/.local/lib/python2.7/site-packages/spynnaker/pyNN/models/neural_projections/connectors/distance_dependent_probability_connector.pyc in __init__(self, d_expression, allow_self_connections, safe, verbose, n_connections)
     56         # TODO: Work out how this can be done statistically
     57         expand_distances = self._expand_distances(self._d_expression)
---> 58         pre_positions = self._pre_population.positions
     59         post_positions = self._post_population.positions
     60 

AttributeError: 'NoneType' object has no attribute 'positions'
```

the file causing this issue is located at
<your-python2.7-systemfolder>/spynnaker/pyNN/models/neural_projections/connectors/distance_dependent_probability_connector.py (Line 58)

I am not quite familiar with the pyNN architecture but I assume its impossible for the connector to know already on construction which are the pre and post populations will be.

```
import spynnaker8 as p
from pyNN.random import RandomDistribution
import pylab
import numpy
import itertools
from pyNN.utility.plotting import Figure, Panel

p.setup(timestep=1.0)

def create_grid(n, label, dx=1., dy=1.):
    grid_structure = p.Grid2D(dx=dx, dy=dy, x0=0, y0=0)
    return p.Population(n*n, p.IF_curr_exp(), structure=grid_structure,
                        label=label)

n = 10
tau_exc = 1.
tau_inh = 1.

# Network
grid = create_grid(n, n)

# Connectors
dist_dep_exc = 'exp(-abs(d)/{tau_exc})'.format(tau_exc=tau_exc)
dist_dep_inh = 'exp(-abs(-d)/{tau_inh})'.format(tau_inh=tau_inh)

import ipdb
# ipdb.set_trace()

exc_connector = p.DistanceDependentProbabilityConnector(dist_dep_exc, allow_self_connections=False)
                                                        # weights=5.,
                                                        # delays='abs(d)')

inh_connector = p.DistanceDependentProbabilityConnector(dist_dep_inh, allow_self_connections=False)
                                                        # weights=5.,
                                                        # delays=0.1)
# Wire grid
p.Projection(grid, grid, exc_connector)
p.Projection(grid, grid, inh_connector)

p.run(10000)
```

**Comment by andrewgait** [comment]
The same problem is also present with the SmallWorldConnector as well.

**Comment by andrewgait** [comment]
This should now be working for both PyNN 0.7 and PyNN 0.8, with one caveat - in PyNN 0.8, you need to edit Grid2D.calculate_size(...) in pyNN/space.py to 

`    def calculate_size(self, n):
        """docstring goes here"""
        nx = math.sqrt(n * self.aspect_ratio)
        if n % nx != 0:
            raise Exception("Invalid size: n=%g, nx=%d" % (n, nx))
#        ny = n / nx
        nx = int(round(nx))
        ny = n // nx
        return nx, ny`

otherwise it won't work.

This is a change that has been made between PyNN 0.8 and 0.9; I believe if you use the scripts for sPyNNaker8 with PyNN 0.9 then they should work, but I haven't yet tested that.

**Comment by andrewgait** [comment]
Adding a note here too: this change is, as far as I can tell, only on the current master branch of PyNN and not in any released version (certainly not in 0.9.1, which I've just downloaded); so you'll have to do the edit of the PyNN 0.8 source code yourself for now if you want to use these connectors.

**Comment by andrewgait** [comment]
This has been fixed since PyNN 0.9.2, so closing this issue.
